sexpress/src/main/scala/org/ensime/sexp/SexpParser.scala

```diff
@@ -1,122 +1,166 @@
-    oneOrMore(Alpha | Digit | SymbolSpecial) ~ zeroOrMore(Alpha | Digit | SymbolSpecial | ".") ~ optional("?")
-  } ~> { sym =>
-    if (sym == "nil") SexpNil
-    else SexpSymbol(sym)
+  def Digits = rule {
+    oneOrMore(Digit)
+  }
+
+  def Frac = rule {
+    "." ~ Digits
   }
 
-  private lazy val SexpEmptyList: Rule1[SexpNil.type] = rule("List(empty)") {
-    (LeftBrace ~ RightBrace)
-  } ~> { _ => SexpNil }
+  def Exp = rule {
+    ignoreCase('e') ~ optional(anyOf("+-")) ~ Digits
+  }
 
-  private lazy val SexpListP: Rule1[Sexp] = rule("List") {
-    LeftBrace ~ (SexpP ~ zeroOrMore(Whitespace ~ SexpP)) ~ RightBrace
-  } ~~> {
-    (head, tail) => SexpList(head :: tail)
+  private def SexpNaNP: Rule1[SexpAtom] = rule {
+    "-1.0e+INF" ~ push(SexpNegInf) |
+      "1.0e+INF" ~ push(SexpPosInf) |
+      optional('-') ~ "0.0e+NaN" ~ push(SexpNaN)
   }
 
-  private lazy val SexpConsP: Rule1[SexpCons] = rule("Cons") {
-    LeftBrace ~
-      SexpP ~ Whitespace ~ "." ~ Whitespace ~ SexpP ~ RightBrace
-  } ~~> {
-    (x, y) => SexpCons(x, y)
+  private def SexpQuotedP: Rule1[Sexp] = rule {
+    "'" ~ SexpP ~> { v: Sexp => SexpCons(SexpQuote, v) }
+  }
+
+  private def SexpSymbolP: Rule1[SexpAtom] = rule {
+    // ? allowed at the end of symbol names
+    capture(oneOrMore(Alpha | Digit | SymbolSpecial) ~ zeroOrMore(Alpha | Digit | SymbolSpecial | ".") ~ optional("?")) ~> { sym: String =>
+      if (sym == "nil") SexpNil
+      else SexpSymbol(sym)
+    }
+  }
+
+  private def SexpEmptyList: Rule1[SexpNil.type] = rule {
+    LeftBrace ~ RightBrace ~ push(SexpNil)
   }
 
   private val SexpQuote = SexpSymbol("quote")
-  private lazy val SexpQuotedP: Rule1[Sexp] = rule("Quoted") {
-    "'" ~ SexpP
-  } ~~> { v => SexpCons(SexpQuote, v) }
 
-  private lazy val Character: Rule1[String] = rule("Character") { EscapedChar | NormalChar }
-  private lazy val EscapedChar: Rule1[String] = rule("EscapedChar") {
-    "\\" ~ ANY ~> { s => unescape(s) }
+  private def Character: Rule1[String] = rule {
+    EscapedChar | NormalChar
   }
-  private lazy val NormalChar: Rule1[String] = rule("NormalChar") { !anyOf("\"\\") ~ ANY ~> identity }
 
-  // Rule0 primitives and helpers...
-  private lazy val Alpha = rule("Alpha") { "a" - "z" | "A" - "Z" }
-  private lazy val Integer = rule("Integer") { optional("-") ~ (("1" - "9") ~ Digits | Digit) }
-  private lazy val Digits = rule("Digits") { oneOrMore(Digit) }
-  private lazy val Digit = rule("Digit") { "0" - "9" }
-  private lazy val Frac = rule("Frac") { "." ~ Digits }
-  private lazy val Exp = rule("Exp") { ignoreCase("e") ~ optional(anyOf("+-")) ~ Digits }
+  private def EscapedChar: Rule1[String] = rule {
+    '\\' ~ capture(ANY) ~> { s: String => unescape(s) }
+  }
 
-  private lazy val SymbolSpecial = rule("SymbolSpecial") { anyOf("+-*/_~!@$%^&=:<>{}") }
+  private def NormalChar: Rule1[String] = rule {
+    capture(CharPredicate.Printable -- "\"\\") ~> { s: String => s }
```

Comment:
This is not great.
The `--` recreates a new CharPredicate here on *every rule application*!
You should factor this out and put it into the companion.

Also, there is no need for `capture` here. Just use `lastChar` in the action.
This also makes `NormalChar` a `Rule1[Char]` instead of `Rule1[String]`, which appears the more natural choice.